### PR TITLE
Vectorize connections

### PR DESCRIPTION
**Pull request by alexfikl**

### CODE REVIEW

File: meshmode/discretization/connection/chained.py [hunk 1]
```diff
@@ -61,11 +62,12 @@ def __init__(self, connections, from_discr=None):
 
         self.connections = connections
 
-    def __call__(self, vec):
+    @obj_array_vectorized_n_args
```
Comment (inducer) [inline meshmode/discretization/connection/chained.py:65]:
```suggestion
    @obj_array_vectorized

File: meshmode/discretization/connection/direct.py [hunk 2]
```diff
@@ -249,11 +250,15 @@ def full_resample_matrix(self, actx):
 
         return make_direct_full_resample_matrix(actx, self)
 
+    @obj_array_vectorized_n_args
```
Comment (inducer) [inline meshmode/discretization/connection/direct.py:253]:
```suggestion
    @obj_array_vectorized
```

File: meshmode/discretization/connection/direct.py [hunk 3]
```diff
@@ -249,11 +250,15 @@ def full_resample_matrix(self, actx):
...
     def __call__(self, ary):
         from meshmode.dof_array import DOFArray
         if not isinstance(ary, DOFArray):
             raise TypeError("non-array passed to discretization connection")
 
+        if ary.shape != (len(self.from_discr.groups),):
+            raise ValueError("array must have same size as discretization groups")
```
Comment (inducer) [inline meshmode/discretization/connection/direct.py:260]: Rephrase

File: meshmode/discretization/connection/projection.py [hunk 4]
```diff
@@ -115,11 +116,15 @@ def det(v):
...
-    def __call__(self, vec):
-        if not isinstance(vec, DOFArray):
+    @obj_array_vectorized_n_args
+    def __call__(self, ary):
+        if not isinstance(ary, DOFArray):
             raise TypeError("non-array passed to discretization connection")
 
-        actx = vec.array_context
+        if ary.shape != (len(self.from_discr.groups),):
+            raise ValueError("array must have same size as discretization groups")
```
Comment (inducer) [inline meshmode/discretization/connection/projection.py:125]: Rephrase

File: meshmode/discretization/connection/projection.py [hunk 5]
```diff
@@ -115,11 +116,15 @@ def det(v):
 
         return weights
 
-    def __call__(self, vec):
-        if not isinstance(vec, DOFArray):
+    @obj_array_vectorized_n_args
```
Comment (inducer) [inline meshmode/discretization/connection/projection.py:119]:
```suggestion
    @obj_array_vectorized
```

**Review by inducer — COMMENTED** [review]
Thanks!

File: meshmode/discretization/connection/chained.py (hunk 1 above)
Comment (alexfikl) [reply meshmode/discretization/connection/chained.py:65]: That didn't seem to work. I guess because it passed `self` in too?

Comment (inducer) [reply meshmode/discretization/connection/chained.py:65]: Ah, fair. Nvm al these then.

File: meshmode/discretization/connection/direct.py (hunk 3 above)
Comment (alexfikl) [reply meshmode/discretization/connection/direct.py:260]: Hm.. "array size does not match incomming discretization"? "expected array of size %d, but got %d"?

**Comment by inducer** [comment]
Either is fine.

File: meshmode/discretization/connection/direct.py (hunk 3 above)
Comment (alexfikl) [reply meshmode/discretization/connection/direct.py]: Should be fixed in f4a0385.

Actually realized that the `DirectDiscretizationConnection` already had this check lower down, so just moved it up before the allocations.

**Comment by inducer** [comment]
Thx!

**Pull request merged**